### PR TITLE
Add 'fields' parameter to ProxySpace to Conditions

### LINKED ISSUE #236: Implement API to pass 'fields' parameter to CRUD library
В java любая crud операция возвращает всегда все поля (весь tuple целиком). в crud есть возможность задать в opts столбцы (fields), которые нужно вернуть, но в cartridge-driver 0.7.2 это задать нельзя, opts там хардкодится, единственное, что ставится, это таймаут.  нельзя ли ликвидировать "разрыв" в функциональности между java и crud ?

к примеру, в кортеже 30 полей, и я не могу выбрать одно поле, только все 30. 
имеется в виду:

space = tntClient.getClient().space(SPACE_NAME);
space.select(Conditions.any()) — возвращает все поля, нет возможности задать нужное (ые), condition для примера

тоже  самое replace/update/upsert/delete

гоняют по сети впустую данные

### PR DESCRIPTION
**Pull request by ArtDu**
What has been done? Why? What problem is being solved?

I didn't forget about

- [ ] Tests
- [ ] Changelog
- [ ] Documentation
- [ ] Cleanup the code for review. See [checklist](https://github.com/tarantool/cartridge-java/blob/master/docs/review-checklist.md)

Closes #236

### CODE REVIEW

File: src/main/java/io/tarantool/driver/core/space/ProxyTarantoolSpace.java [hunk 1]
```diff
@@ -207,12 +245,12 @@ private CompletableFuture<R> upsert(Conditions conditions,
     public CompletableFuture<Void> truncate() throws TarantoolClientException {
         try {
             return executeVoidOperation(TruncateProxyOperation.<Void>builder()
-                    .withClient(client)
-                    .withSpaceName(spaceName)
-                    .withFunctionName(operationsMapping.getTruncateFunctionName())
-                    .withRequestTimeout(config.getRequestTimeout())
-                    .build()
-            );
+                                                .withClient(client)
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/core/space/ProxyTarantoolSpace.java:248]: Extra spacing?

File: src/main/java/io/tarantool/driver/core/proxy/AbstractProxyOperation.java [hunk 2]
```diff
@@ -55,4 +56,84 @@ public CallResultMapper<T, SingleValueCallResult<T>> getResultMapper() {
...
+         *
+         * @param functionName name of Tarantool server function
+         * @return builder
+         */
+        public B withFunctionName(String functionName) {
+            this.functionName = functionName;
+            return self();
+        }
+
+        public B withArgumentsMapper(MessagePackObjectMapper objectMapper) {
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/core/proxy/AbstractProxyOperation.java:107]: These two methods are missing a Javadoc

**Review by akudiyar — CHANGES_REQUESTED** [review]
I'm not sure about putting `fields` into `Conditions`. This class was intended and can be used only for select operations. Using it for the other operations that cannot make use of other `Conditions` fields leads to user confusion and we will have to add a lot of extra checks that the user haven't specified any unsupported fields.

 I see two options here:
 - making a new class that is used as base class for Conditions and can be passed to all operations, so it does not allow using extra fields where they aren't supported
 - judging by the output type about what fields should be included. This approach can be combined with the first one or implemented independently. But to me it looks closer to what users may expect and eliminate any confusion that may arise from the difference between the passed field names and the available output type fields. However, matching fields to the output type may be handed over to the Spring Data connector, as it has more control over the output structures.
 
 BTW, I like the ProxyOperations refactoring, wouldn't it be better to merge it separately first?

File: src/main/java/io/tarantool/driver/core/proxy/AbstractProxyOperation.java (hunk 2 above)
Comment (ArtDu) [reply src/main/java/io/tarantool/driver/core/proxy/AbstractProxyOperation.java:107]: Fixed and exported to https://github.com/tarantool/cartridge-java/pull/256

**Comment by ArtDu** [comment]
> I'm not sure about putting `fields` into `Conditions`. This class was intended and can be used only for select operations. Using it for the other operations that cannot make use of other `Conditions` fields leads to user confusion and we will have to add a lot of extra checks that the user haven't specified any unsupported fields.
> 
> I see two options here:
> 

> * judging by the output type about what fields should be included. This approach can be combined with the first one or implemented independently. But to me it looks closer to what users may expect and eliminate any confusion that may arise from the difference between the passed field names and the available output type fields. However, matching fields to the output type may be handed over to the Spring Data connector, as it has more control over the output structures.
> 
> BTW, I like the ProxyOperations refactoring, wouldn't it be better to merge it separately first?

I agree with you that using `Condition` is not the best solution. It's interesting about seeing in output type, but we need a more general way to pass not only `fields` but other parameters in crud(it's possible we add this option later). Maybe we add another class - `Option` and will inherit it for specific methods options? I don't see how to use `Condition` and base parent class, because we will still have all `select` options in Condition and it will be confusing

**Comment by akudiyar** [comment]
The fact that any other methods are using Conditions now even if they aren't able to use all options there needs to be fixed, I think.
An option that I see here -- to add an `OperationOptions` base class, that will contain base options like timeouts, pass it to every method, and for CRUD/Proxy space/client implementation allow to pass a child class there that supports more options, only specific to the cluster client.
That means we will leave the existing method signatures as is, and add new variations supporting this new `OperationOption` as a parameter. For the single node implementation, it will support only the base set of options (like timeouts only). Thus we will not break the single server compatibility.

Examples:
 - Single node client:
```java
   Conditions conditions = Conditions.any();
   OperationOptions options = BaseOperationOptions.builder().withTimeout(100500).build();
   client.space("test").select(conditions, options);
``` 
 - Cluster client:
 ```java
   Conditions conditions = Conditions.any();
   OperationOptions options = BaseOperationOptions.builder()
       .withTimeout(100500)
       .withFields(Arrays.asList("id", "name"))
       .build();
   client.space("test").select(conditions, options);
 ```
 
 Notes:
  - we'll have to have casts in the cluster space implementation in the naive solution. The other (better) option is to parameterize the space and client classes with the options class, sigh... (I hate Java)
  - we'll have to generate another dozen of methods with parameter combinations for this new parameter for better UX (I hate Java).
  - at the moment we are not passing the `batchSize` option correctly to tarantool/crud because we don't have the same option in the standalone server (we pass the `Conditions.limit()` value to that). It will be better to have this option in the new class. However, it's not right for other options like `startAfter`, because they are related to the select operation semantics and not to the internal tarantool/crud library tweaks.
  - the new options related to batch operations will also be supported by this new options class.

**Pull request closed without merging**